api/client/run.go

```diff
@@ -105,154 +103,152 @@ func (capi ClientApi) Run(conf map[string]string, testrunName string, displayRep
-		fmt.Print("RUNNING TEST: ", testUuid)
-		fmt.Print("\n\n")
+	err = listenForTestStatus(conf)
+	if err != nil {
+		fmt.Println(err, "Will now watch the testrun metrics API for 45 seconds to see if we get a result that way. Please wait...")
+	}
 
-		fmt.Println("(Please be patient while the test finishes...)\n")
+	// Poll for results
+	timeout := time.After(45 * time.Second)
+	data, err := getRunResult(conf, testUUID)
+	for err != nil {
+		select {
+		case <-timeout:
+			fmt.Println("Failed to retrieve test data after 45 seconds. Something must be wrong - quitting.")
+			os.Exit(1)
+		default:
+			time.Sleep(1 * time.Second)
+			data, err = getRunResult(conf, testUUID)
+		}
+	}
 
-		retries := 0
-	retry:
-		// connect to this socket
-		conn, err := net.Dial("tcp", fmt.Sprintf("%s:8081", conf["host"]))
-		if err != nil {
+	fmt.Printf("\n\nDone.\n")
 
-			retries = retries + 1
-			time.Sleep(1000 * time.Millisecond)
-			if retries > 5 {
-				fmt.Println("Failed to subscribe to test event stream after several retries.")
-				fmt.Println("Will now watch the testrun metrics API for 45 seconds to see if we get a result that way. Please wait...")
-				goto tcpfailed
-			} else {
-				fmt.Println("Failed to subscribe to test event stream. Retrying...")
-				goto retry
-			}
+	// display it to the user if desired
+	if sourceGroup != "" || displayReport {
+		var buf bytes.Buffer
+		err := json.Indent(&buf, data, "", "  ")
+		if err != nil {
+			fmt.Printf("error %q: %v\n", string(data), err)
 		}
+		buf.WriteTo(os.Stdout)
+		fmt.Println()
+	}
 
-		for {
+	os.Exit(0)
+}
 
-			// listen for reply
-			message, err := bufio.NewReader(conn).ReadString('\n')
+var errNoTestResult = errors.New("No test result")
 
-			// If an error is raised, it's probably because the server killed the connection
-			if err != nil {
-				// TODO(mierdin): This doesn't really tell us if the connection died because of an error or not
-				break
-			}
+// getRunResult collects the results of test run from the server's REST API
+func getRunResult(conf map[string]string, testUUID string) ([]byte, error) {
+	// Go back and get our testrun data
+	url := fmt.Sprintf("http://%s:%s/v1/testdata?testUuid=%s", conf["host"], conf["port"], testUUID)
 
-			var statuses map[string]string
-			err = json.Unmarshal([]byte(message), &statuses)
-			// TODO (mierdin): Handle error
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
 
-			if firstMessage == false {
-				recordCount = len(statuses)
-				firstMessage = true
-			}
+	// Defer the closing of the body
+	defer resp.Body.Close()
 
-			init, ready, testing, finished := 0, 0, 0, 0
-			for _, status := range statuses {
-
-				switch status {
-				case "init":
-					init += 1
-				case "ready":
-					ready++
-				case "testing":
-					testing++
-				case "finished":
-					finished++
-				default:
-					fmt.Println("Invalid status recieved.")
-					os.Exit(1)
-				}
-			}
+	if resp.StatusCode > 299 {
+		return nil, errNoTestResult
```

Comment:
Okay, I can get on board with custom errors where appropriate. I do agree that it is silly to have the consumer parse the error message in order to know what to do with it, so I would also like to add that I'm okay with the ```errors.New()``` syntax insofar as that error won't NEED to be parsed. In this particular example, the consumer doesn't care what the message is, what the error type is, etc. It just cares that the err != nil. In cases like this, I don't care as much about one thing vs the other, since there's really not much of a decision to be made there.